File: app/src/main/java/workshop/akbolatss/tagsnews/screen/sources/SourcesActivity.java
package workshop.akbolatss.tagsnews.screen.sources;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.PopupMenu;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.github.clans.fab.FloatingActionButton;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.OnClick;
import workshop.akbolatss.tagsnews.R;
import workshop.akbolatss.tagsnews.base.BaseActivity;
import workshop.akbolatss.tagsnews.di.component.DaggerSourcesComponent;
import workshop.akbolatss.tagsnews.di.module.SourcesModule;
import workshop.akbolatss.tagsnews.repositories.source.RssSource;
import workshop.akbolatss.tagsnews.screen.sources.helper.SimpleItemTouchHelperCallback;

public class SourcesActivity extends BaseActivity implements SourcesView, SourcesAdapter.OnRssClickListener {

    @Inject
    protected Context mContext;

    @Inject
    protected SourcesPresenter mPresenter;

    @BindView(R.id.btnFabAdd)
    protected FloatingActionButton btnFabAdd;

    @BindView(R.id.toolbar)
    protected Toolbar mToolbar;

    @BindView(R.id.tvNoContent)
    protected TextView mNoContent;

    @BindView(R.id.progressBar)
    protected ProgressBar mProgressBar;

    @BindView(R.id.recyclerView)
    protected RecyclerView mRecyclerView;
    private SourcesAdapter mSourcesAdapter;

    @Override
    protected void onViewReady(Bundle savedInstanceState, Intent intent) {
        super.onViewReady(savedInstanceState, intent);
        setSupportActionBar(mToolbar);
        getSupportActionBar().setDisplayShowTitleEnabled(false);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        DaggerSourcesComponent.builder()
                .appComponent(getAppComponent())
                .sourcesModule(new SourcesModule(this))
                .build()
                .inject(this);

        onInitRecycler();

        mPresenter.onLoadSources();
    }

    private void onInitRecycler() {
        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(mContext, LinearLayoutManager.VERTICAL, false));
        mRecyclerView.setNestedScrollingEnabled(false);

        mSourcesAdapter = new SourcesAdapter(this);
        mRecyclerView.setAdapter(mSourcesAdapter);

        ItemTouchHelper.Callback callback = new SimpleItemTouchHelperCallback(mSourcesAdapter);
        ItemTouchHelper touchHelper = new ItemTouchHelper(callback);
        touchHelper.attachToRecyclerView(mRecyclerView);
    }

    @OnClick(R.id.btnFabAdd)
    @Override
    public void onAddNewSource() {
        LayoutInflater layoutInflater = LayoutInflater.from(SourcesActivity.this);
        final View subView = layoutInflater.inflate(R.layout.dialog_new_source, null);
        final EditText etLink = subView.findViewById(R.id.etLink);

        AlertDialog.Builder builder = new AlertDialog.Builder(this, R.style.Dialog);

        builder.setTitle(R.string.tvEnter);
        builder.setView(subView);
        builder.setPositiveButton(R.string.tvAdd, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                RssSource rssSource = new RssSource();
                rssSource.setIsActive(true);
                rssSource.setTitle(etLink.getText().toString());
                rssSource.setLink(etLink.getText().toString());
                mPresenter.onAddNewSource(rssSource);

                mSourcesAdapter.onAddItem(rssSource);
                dialogInterface.dismiss();
            }
        });
        builder.setNegativeButton(R.string.tvCancel, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                dialogInterface.cancel();
            }
        });
        builder.show();
    }

    @Override
    public void onLoadSources(List<RssSource> rssSourceList) {
        mSourcesAdapter.onAddItems(rssSourceList);
    }

    @Override
    public void onItemsSwapped(RssSource from, RssSource to) {
        mPresenter.onSwapPositions(from, to);
    }

    @Override
    public void onSourceOptions(final RssSource rssSource, View view, final int pos) {
        PopupMenu popupMenu = new PopupMenu(this, view);
        popupMenu.inflate(R.menu.menu_popup_source);
        popupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                switch (item.getItemId()) {
                    case R.id.mRename:
                        LayoutInflater layoutInflater = LayoutInflater.from(SourcesActivity.this);
                        final View subView = layoutInflater.inflate(R.layout.dialog_new_source, null);
                        final EditText etLink = subView.findViewById(R.id.etLink);

                        etLink.setText(rssSource.getTitle());
                        etLink.setHint(R.string.tvNewName);
                        AlertDialog.Builder builder = new AlertDialog.Builder(SourcesActivity.this, R.style.Dialog);

                        builder.setTitle(R.string.tvEnter);
                        builder.setView(subView);
                        builder.setPositiveButton(R.string.tvEdit, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                                rssSource.setTitle(etLink.getText().toString());
                                mPresenter.onUpdateSource(rssSource);
                                mSourcesAdapter.onUpdateItem(rssSource, pos);
                                dialogInterface.dismiss();
                            }
                        });
                        builder.setNegativeButton(R.string.tvCancel, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                                dialogInterface.cancel();
                            }
                        });
                        builder.show();
                        return true;
                    case R.id.mReaddress:
                        LayoutInflater layoutInflater2 = LayoutInflater.from(SourcesActivity.this);
                        final View subView2 = layoutInflater2.inflate(R.layout.dialog_new_source, null);
                        final EditText etLink2 = subView2.findViewById(R.id.etLink);

                        etLink2.setText(rssSource.getLink());
                        etLink2.setHint(R.string.tvNewAddress);
                        AlertDialog.Builder builder2 = new AlertDialog.Builder(SourcesActivity.this, R.style.Dialog);

                        builder2.setTitle(R.string.tvEnter);
                        builder2.setView(subView2);
                        builder2.setPositiveButton(R.string.tvEdit, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                                rssSource.setLink(etLink2.getText().toString());
                                mPresenter.onUpdateSource(rssSource);
                                mSourcesAdapter.onUpdateItem(rssSource, pos);
                                dialogInterface.dismiss();
                            }
                        });
                        builder2.setNegativeButton(R.string.tvCancel, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                                dialogInterface.cancel();
                            }
                        });
                        builder2.show();
                        return true;
                    case R.id.mRemove:
                        mPresenter.onRemoveSource(rssSource);
                        mSourcesAdapter.onRemoveItem(pos);
                        return true;
                }
                return false;
            }
        });
        popupMenu.show();
    }

    @Override
    public void onSourceSwitch(RssSource rssSource, boolean isActivated, int pos) {
        if (isActivated) {
            rssSource.setIsActive(true);
        } else {
            rssSource.setIsActive(false);
        }
        mPresenter.onUpdateSource(rssSource);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    protected int getContentView() {
        return R.layout.activity_sources;
    }

    @Override
    public void onShowLoading() {
        mRecyclerView.setVisibility(View.GONE);
        mProgressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void onHideLoading() {
        mRecyclerView.setVisibility(View.VISIBLE);
        mProgressBar.setVisibility(View.GONE);
    }

    @Override
    public void onNoContent(boolean isEmpty) {
        if (isEmpty) {
            mRecyclerView.setVisibility(View.GONE);
            mNoContent.setVisibility(View.VISIBLE);
        } else {
            mRecyclerView.setVisibility(View.VISIBLE);
            mNoContent.setVisibility(View.GONE);
        }
    }

    @Override
    public void onUnknownError(String errorMessage) {
        Toast.makeText(mContext, R.string.unknown_error, Toast.LENGTH_LONG).show();
    }

    @Override
    public void onTimeout() {
        Toast.makeText(mContext, R.string.timeout_error, Toast.LENGTH_LONG).show();
    }

    @Override
    public void onNetworkError() {
        Toast.makeText(mContext, R.string.network_error, Toast.LENGTH_LONG).show();
    }
}

File: app/src/main/java/workshop/akbolatss/tagsnews/screen/sources/SourcesActivity.kt
package workshop.akbolatss.tagsnews.screen.sources

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.PopupMenu
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.widget.EditText
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_sources.*
import workshop.akbolatss.tagsnews.R
import workshop.akbolatss.tagsnews.base.BaseActivity
import workshop.akbolatss.tagsnews.di.component.DaggerSourcesComponent
import workshop.akbolatss.tagsnews.di.module.SourcesModule
import workshop.akbolatss.tagsnews.model.dao.RssSource
import workshop.akbolatss.tagsnews.screen.sources.helper.SimpleItemTouchHelperCallback
import javax.inject.Inject

/**
 * Activity for managing RSS sources, e.g. add, edit, remove
 */
class SourcesActivity : BaseActivity(), SourcesView, SourcesAdapter.SourceListener {

    @Inject
    lateinit var mContext: Context

    @Inject
    lateinit var mPresenter: SourcesPresenter

    private var mSourcesAdapter: SourcesAdapter? = null
    private var mItemTouchHelper: ItemTouchHelper? = null

    override fun onViewReady(savedInstanceState: Bundle?, intent: Intent) {
        super.onViewReady(savedInstanceState, intent)
        initToolbar()
        initRV()
        initListeners()
        mPresenter.loadSources()
    }

    private fun initToolbar() {
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayShowTitleEnabled(false)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
    }

    private fun initListeners() {
        fabAdd.setOnClickListener {
            onAddNewSource()
        }
    }

    private fun initRV() {
        recyclerView.setHasFixedSize(true)
        recyclerView.layoutManager = LinearLayoutManager(mContext, LinearLayoutManager.VERTICAL, false)

        mSourcesAdapter = SourcesAdapter(this)
        recyclerView.adapter = mSourcesAdapter

        val callback = SimpleItemTouchHelperCallback(mSourcesAdapter)
        mItemTouchHelper = ItemTouchHelper(callback)
        mItemTouchHelper!!.attachToRecyclerView(recyclerView)
    }

    override fun onStartDrag(viewHolder: RecyclerView.ViewHolder) {
        mItemTouchHelper!!.startDrag(viewHolder)
    }

    /**
     * Add manually new RSS source
     */
    override fun onAddNewSource() {
        val layoutInflater = LayoutInflater.from(this@SourcesActivity)
        val subView = layoutInflater.inflate(R.layout.dialog_new_source, null)
        val etLink = subView.findViewById<EditText>(R.id.etLink)

        val builder = AlertDialog.Builder(this, R.style.Dialog)

        builder.setTitle(R.string.tvEnter)
        builder.setView(subView)
        builder.setPositiveButton(R.string.tvAdd) { dialogInterface, i ->
            val rssSource = RssSource()
            rssSource.isActive = true
            rssSource.title = etLink.text.toString().trim()
            rssSource.link = etLink.text.toString().trim()
            if (mPresenter.checkInSources(rssSource)) {
                Toast.makeText(this, resources.getString(R.string.rss_exists), Toast.LENGTH_LONG).show()
            } else {
                mPresenter.addNewSource(rssSource)

                mSourcesAdapter!!.onAddItem(rssSource)
            }
            dialogInterface.dismiss()
        }
        builder.setNegativeButton(R.string.tvCancel) { dialogInterface, i -> dialogInterface.cancel() }
        builder.show()
    }

    /**
     * Load list of all RSS sources to Adapter
     */
    override fun onLoadSources(rssSourceList: List<RssSource>) {
        mSourcesAdapter!!.onAddItems(rssSourceList)
    }

    /**
     * Swap RSS items
     */
    override fun onItemsSwapped(from: RssSource, to: RssSource) {
        mPresenter.swapPositions(from, to)
    }

    /**
     * RSS source options. Edit name, edit link and remove
     */
    override fun onSourceOptions(rssSource: RssSource, view: View, pos: Int) {
        val popupMenu = PopupMenu(this, view)
        popupMenu.inflate(R.menu.menu_popup_source)
        popupMenu.setOnMenuItemClickListener(PopupMenu.OnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.mRename -> {
                    val layoutInflater = LayoutInflater.from(this@SourcesActivity)
                    val subView = layoutInflater.inflate(R.layout.dialog_new_source, null)
                    val etLink = subView.findViewById<EditText>(R.id.etLink)

                    etLink.setText(rssSource.title)
                    etLink.setHint(R.string.tvNewName)
                    val builder = AlertDialog.Builder(this@SourcesActivity, R.style.Dialog)

                    builder.setTitle(R.string.tvEnter)
                    builder.setView(subView)
                    builder.setPositiveButton(R.string.tvEdit) { dialogInterface, i ->
                        rssSource.title = etLink.text.toString()
                        mPresenter.updateSource(rssSource)
                        mSourcesAdapter!!.onUpdateItem(rssSource, pos)
                        dialogInterface.dismiss()
                    }
                    builder.setNegativeButton(R.string.tvCancel) { dialogInterface, i -> dialogInterface.cancel() }
                    builder.show()
                    return@OnMenuItemClickListener true
                }
                R.id.mReaddress -> {
                    val layoutInflater2 = LayoutInflater.from(this@SourcesActivity)
                    val subView2 = layoutInflater2.inflate(R.layout.dialog_new_source, null)
                    val etLink2 = subView2.findViewById<EditText>(R.id.etLink)

                    etLink2.setText(rssSource.link)
                    etLink2.setHint(R.string.tvNewAddress)
                    val builder2 = AlertDialog.Builder(this@SourcesActivity, R.style.Dialog)

                    builder2.setTitle(R.string.tvEnter)
                    builder2.setView(subView2)
                    builder2.setPositiveButton(R.string.tvEdit) { dialogInterface, i ->
                        rssSource.link = etLink2.text.toString()
                        mPresenter.updateSource(rssSource)
                        mSourcesAdapter!!.onUpdateItem(rssSource, pos)
                        dialogInterface.dismiss()
                    }
                    builder2.setNegativeButton(R.string.tvCancel) { dialogInterface, i -> dialogInterface.cancel() }
                    builder2.show()
                    return@OnMenuItemClickListener true
                }
                R.id.mRemove -> {
                    mPresenter.removeSource(rssSource)
                    mSourcesAdapter!!.onRemoveItem(pos)
                    return@OnMenuItemClickListener true
                }
            }
            false
        })
        popupMenu.show()
    }

    /**
     * Activate/deactivate RSS source in #BoardActivity
     * @see BoardActivity
     */
    override fun onSourceSwitch(rssSource: RssSource, isActive: Boolean, pos: Int) {
        rssSource.isActive = isActive
        mPresenter.updateSource(rssSource)
    }

    /**
     * Show loading state
     */
    override fun onShowLoading() {
        recyclerView.visibility = View.GONE
        progressBar.visibility = View.VISIBLE
    }

    /**
     * Hide loading state
     */
    override fun onHideLoading() {
        recyclerView.visibility = View.VISIBLE
        progressBar.visibility = View.GONE
    }

    /**
     * Show if there no content and vice-versa
     */
    override fun onNoContent(isEmpty: Boolean) {
        if (isEmpty) {
            recyclerView.visibility = View.GONE
            tvNoContent.visibility = View.VISIBLE
        } else {
            recyclerView.visibility = View.VISIBLE
            tvNoContent.visibility = View.GONE
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onInitDagger() {
        DaggerSourcesComponent.builder()
                .appComponent(appComponent)
                .sourcesModule(SourcesModule(this))
                .build()
                .inject(this)
    }

    override fun getContentView(): Int {
        return R.layout.activity_sources
    }

    override fun onUnknownError(errorMessage: String) {
        Toast.makeText(mContext, R.string.unknown_error, Toast.LENGTH_LONG).show()
    }

    override fun onTimeout() {
        Toast.makeText(mContext, R.string.timeout_error, Toast.LENGTH_LONG).show()
    }

    override fun onNetworkError() {
        Toast.makeText(mContext, R.string.network_error, Toast.LENGTH_LONG).show()
    }


}
